### PR TITLE
Fix transposed convolutions

### PR DESCRIPTION
**Pull request by atamazov**
This fixes issue #2459. Reasons of the problem:
- The immediate reason is some mistake during refactoring (PR #2303 is missing necessary changes in convolution_api.cpp).
- The root reason is missing test for asymmetric (C != K) transposed convolutions.

Overview:
- Issue fixed (see 12f2c9d21)
- Test added. It also serves as a regression test for #2459.
- By-products:
  - [quality] Resolved review comment: https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2125/files#r1372100824
  - [quality] Refactored IsValidFilterChannelNumber
  - [debugging] Fixed printing of driver commands for transposed convolutions.
  - [driver] Adjusted tolerance for transposed FP32 WrW convolutions.

## :cyclone: Testing results

Preconditions:
- Navi21 (gfx1030/36)
- ROCm 5.7, Ubuntu 20.04.5 LTS
- BUILD_DEV=On, Release build

What tested:
- 93 FP32 transposed convolution configs
- 93 FP16 transposed convolution configs
- Tested modes: DYNAMIC_HYBRID Find mode and Immediate mode.

The list of configs, detailed logs & csv files are available upon request.

----
[Attribution] @junliume @JehandadKhan
- https://github.com/ROCmSoftwarePlatform/MIOpen/labels/bug
- https://github.com/ROCmSoftwarePlatform/MIOpen/labels/urgency_blocker
- Proposed reviewers:
  - @JehandadKhan
  - @carlushuang
  - @iq136boy
  - @muralinr 
  - @averinevg

### CODE REVIEW

File: driver/conv_driver.hpp [hunk 1]
```diff
@@ -1256,6 +1240,8 @@ int ConvDriver<Tgpu, Tref>::AllocateBuffersAndCopy()
     size_t wei_sz = GetTensorSize(weightTensor);
     size_t out_sz = GetTensorSize(outputTensor);
     auto subnorm_percentage = miopen::Value(MIOPEN_DRIVER_SUBNORM_PERCENTAGE{});
+    if(subnorm_percentage != 0)
+        std::cout << "MIOPEN_DRIVER_SUBNORM_PERCENTAGE = " << subnorm_percentage << std::endl;
```
Comment (atamazov) [inline driver/conv_driver.hpp:1244]: This is for @muralinr

File: driver/conv_driver.hpp [hunk 2]
```diff
@@ -3472,8 +3458,7 @@ int ConvDriver<Tgpu, Tref>::VerifyForward()
     }
 
     std::cout << "Forward Convolution Verifies OK on " << (UseGPUReference() ? "GPU" : "CPU")
-              << " reference (" << miopen::Value(MIOPEN_DRIVER_SUBNORM_PERCENTAGE{}) << ", "
-              << " reference (" << error << ')' << std::endl;
+              << " reference (" << error << " < " << tolerance << ')' << std::endl;
```
Comment (atamazov) [inline driver/conv_driver.hpp:3461]: Ditto

File: src/convolution_api.cpp [hunk 3]
```diff
@@ -56,16 +56,13 @@ static inline auto MakeFwdCtxAndProblem(miopenHandle_t handle,
...
-                                                                       miopen::deref(xDesc),
-                                                                       conv,
-                                                                       direction};
+    /// \anchor transpose_convolutions_x_y_swapping
+    /// In transpose mode we exchange x with y. From the other hand, when Backward*
+    /// ProblemDescription instances are constructed, x and y shall be swapped as well.
+    /// As transpose mode swaps Forward with Backward AND x with y, the order of
+    /// ctor arguments remains the same.
+    auto problem = ProblemDescription{
+        miopen::deref(xDesc), miopen::deref(wDesc), miopen::deref(yDesc), conv, direction};
```
Comment (atamazov) [inline src/convolution_api.cpp:65]: The fix part 1

File: src/convolution_api.cpp [hunk 4]
```diff
@@ -82,16 +79,9 @@ static inline auto MakeBwdCtxAndProblem(miopenHandle_t handle,
...
-                                                                       conv,
-                                                                       direction}
-                                                  : ProblemDescription{miopen::deref(dxDesc),
-                                                                       miopen::deref(wDesc),
-                                                                       miopen::deref(dyDesc),
-                                                                       conv,
-                                                                       direction};
+    /// \ref transpose_convolutions_x_y_swapping
+    auto problem = ProblemDescription{
+        miopen::deref(dyDesc), miopen::deref(wDesc), miopen::deref(dxDesc), conv, direction};
```
Comment (atamazov) [inline src/convolution_api.cpp:84]: Part 2. This is the main difference from #2476.

Other changes in this file are NFC.

File: src/driver_arguments.cpp [hunk 5]
```diff
@@ -117,21 +117,28 @@ std::string ConvArgsForMIOpenDriver(const miopen::TensorDescriptor& xDesc,
...
-            << " -l " << convDesc.GetConvDilations()[0]
-            << " -j " << convDesc.GetConvDilations()[1]; // clang-format on
+        ss << " -n " << xDesc.GetLengths()[0] //
+           << " -c " << xDesc.GetLengths()[1] //
+           << " -H " << xDesc.GetLengths()[2] //
+           << " -W " << xDesc.GetLengths()[3] //
+           << " -k "
+           << (convDesc.mode == miopenTranspose        //
+                   ? wDesc.GetLengths()[1]             //
+                   : wDesc.GetLengths()[0])            //
```
Comment (atamazov) [inline src/driver_arguments.cpp:133]: @JehandadKhan This resolves https://github.com/ROCmSoftwarePlatform/MIOpen/issues/2459#issuecomment-1776230996

File: src/ocl/convolutionocl.cpp [hunk 6]
```diff
@@ -65,44 +65,70 @@ MIOPEN_DECLARE_ENV_VAR(MIOPEN_DUMP_TENSOR_PATH)
...
+            return false;
+
+        if(layout == miopenTensorNCHW      //
+           || layout == miopenTensorNCHWc4 //
+           || layout == miopenTensorNCHWc8)
+            return w.GetLengths()[0] % groups == 0;
+
+        if(layout == miopenTensorCHWNc4 //
+           || layout == miopenTensorCHWNc8)
+            return w.GetLengths()[3] % groups == 0;
```
Comment (atamazov) [inline src/ocl/convolutionocl.cpp:106]: @iq136boy @JehandadKhan This is related to Q at https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2125/files#r1372100824. It restores the check deleted there.

**Review by atamazov — COMMENTED** [review]
Hints for reviewers.

File: src/convolution_api.cpp (hunk 3 above)
Comment (atamazov) [reply src/convolution_api.cpp:65]: @averinevg @carlushuang FYI

File: src/convolution_api.cpp (hunk 4 above)
Comment (atamazov) [reply src/convolution_api.cpp:84]: @averinevg @carlushuang FYI

File: driver/conv_driver.hpp (hunk 1 above)
Comment (muralinr) [reply driver/conv_driver.hpp:1244]: LGTM. Thanks for adding me

**Comment by atamazov** [comment]
@junliume Tests added. PR description updated.

**Review by JehandadKhan — APPROVED** [review]

**Review by junliume — APPROVED** [review]

**Pull request merged**